cli/internal/runsummary/run_summary.go

```diff
@@ -91,67 +140,86 @@ func (summary *RunSummary) Save(dir turbopath.AbsoluteSystemPath, singlePackage
-	Dependencies           []string                              `json:"dependencies"`
-	Dependents             []string                              `json:"dependents"`
-	ResolvedTaskDefinition *fs.TaskDefinition                    `json:"resolvedTaskDefinition"`
-	ExpandedInputs         map[turbopath.AnchoredUnixPath]string `json:"expandedInputs"`
-	ExpandedOutputs        []turbopath.AnchoredSystemPath        `json:"expandedOutputs"`
-	Framework              string                                `json:"framework"`
-	EnvVars                TaskEnvVarSummary                     `json:"environmentVariables"`
-	Execution              *TaskExecutionSummary                 `json:"execution,omitempty"` // omit when it's not set
-	ExternalDepsHash       string                                `json:"hashOfExternalDependencies"`
-}
+// record sends the summary to the API
+// TODO: make this work for single package tasks
+func (rsm *Meta) record() []error {
+	errs := []error{}
+
+	// Right now we'll send the POST to create the Run and the subsequent task payloads
+	// when everything after all execution is done, but in the future, this first POST request
+	// can happen when the Run actually starts, so we can send updates to Vercel as the tasks progress.
+	runsURL := fmt.Sprintf(runsEndpoint, rsm.spaceID)
+	var runID string
+	payload := newVercelRunCreatePayload(rsm.RunSummary)
+	if startPayload, err := json.Marshal(payload); err == nil {
+		if resp, err := rsm.apiClient.JSONPost(runsURL, startPayload); err != nil {
+			errs = append(errs, err)
+		} else {
+			vercelRunResponse := &vercelRunResponse{}
+			if err := json.Unmarshal(resp, vercelRunResponse); err != nil {
+				errs = append(errs, err)
+			} else {
+				runID = vercelRunResponse.ID
+			}
+		}
+	}
+
+	if runID != "" {
+		rsm.postTaskSummaries(runID)
+
+		if donePayload, err := json.Marshal(newVercelDonePayload()); err == nil {
+			if _, err := rsm.apiClient.JSONPatch(runsURL, donePayload); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
 
-// TaskEnvVarSummary contains the environment variables that impacted a task's hash
-type TaskEnvVarSummary struct {
-	Configured []string `json:"configured"`
-	Inferred   []string `json:"inferred"`
-	Global     []string `json:"global"`
+	if len(errs) > 0 {
+		return errs
+	}
+
+	return nil
 }
 
-// toSinglePackageTask converts a TaskSummary into a singlePackageTaskSummary
-func (ht *TaskSummary) toSinglePackageTask() singlePackageTaskSummary {
-	dependencies := make([]string, len(ht.Dependencies))
-	for i, depencency := range ht.Dependencies {
-		dependencies[i] = util.StripPackageName(depencency)
+func (rsm *Meta) postTaskSummaries(runID string) {
+	// We make at most 8 requests at a time.
+	maxParallelRequests := 8
+	taskSummaries := rsm.RunSummary.Tasks
+	taskCount := len(taskSummaries)
+	taskURL := fmt.Sprintf(tasksEndpoint, rsm.spaceID, runID)
+
+	parallelRequestCount := maxParallelRequests
```

Comment:
We could maybe simplify this slightly by removing this var, at the cost of running up to 7 noop goroutines... up to you

Reply:
This is pretty much a copy paste of @nathanhammond's implementation of parallelizing cache checks: https://github.com/vercel/turbo/blob/674bd4f5456ab030d5feb74301f6a446770bcc5b/cli/internal/run/dry_run.go#L112. I tend to not want to make small changes like this so if there's an opportunity to consolidate, we don't second-guess ourselves about why those changes exist!